### ISSUE TITLE
Secure Boot with RPI3

### ISSUE THREAD

**Issue by Christian56** [OP-TEE/optee_os#2478]
Hi, 
currently im trying to configure OP-TEE to boot in a secure mode. However as im trying to compile the OS, I get the error that  "platform_oid.h" is missing. It is included in tools/cert_create/src/cert.c
As platform_oid.h is not found in the repo Im not sure how to continue. 
In the original tf there was an option to include tbbr_oid.h instead of platform_oid.h. However this doesnt exists in the repository either. 
As the trusted firmware is modified at least a bit (polarssl instead of hash/signature) I thought i might post the question here.

**Comment by Christian56** [comment]
Hi, 
I reopened this issue because I still having troubles in order to start a secure boot. 
Initially i closed the issue because OP-Tee used an [old version of Trusted Firmware](https://github.com/ARM-software/tf-issues/issues/608). The mentioned PRs have been in merged with OPTEE however I still cannot start the current version of OP_TEE as it freezes at this point on RPI3:

```
Flattened Device Tree blob at 01000000
   Booting using the fdt blob at 0x1000000
   reserving fdt memory region: addr=8000000 size=2000000
   reserving fdt memory region: addr=0 size=1000
   Loading Device Tree to 000000003ab34000, end 000000003ab3d7e8 ... OK

Starting kernel ...

```

So i tried the stable version (-b 3.2.0) and merged the PRs with this version based on 

```
> cd .repo/manifests
> git fetch origin refs/pull/119/head && git checkout FETCH_HEAD
> cd ../../
> repo sync -j8
> cd optee_os
> git fetch github refs/pull/2394/head && git checkout FETCH_HEAD
> cd ../build
> git fetch github refs/pull/269/head && git checkout FETCH_HEAD
> make toolchains && make -j4 all
```
However I get a CPIO error saying that there is not enough memeory accessible for the boot partition. I tried it again and now I didnt get any error message and just a number of blocks... which seems kinda strange (also the current version contains more blocks). When i start the system I got the following which seems like the boot partition is still uncomplete:
```
Device 0: unknown device
Waiting for Ethernet connection... unable to connect.
missing environment variable: pxeuuid
missing environment variable: bootfile
Retrieving file: pxelinux.cfg/01-b8-27-eb-8f-c4-dd
```
Long story short: I want to boot OP-TEE in secure boot and I'm wondering how I can get a stable Version of OP-TEE featuring secure boot. Maybe there is a commit i can roll back to or something. :(

**Comment by igoropaniuk** [comment]
Hi @Christian56,

> However I get a CPIO error saying that there is not enough memeory accessible for the boot partition

Why not to extend boot partition? (just simply do re-partitioning following the same steps in `make img-help`, but specify different size for boot partition)

**Comment by igoropaniuk** [comment]
@Christian56,

Both `armstub8.bin` (contains ATF with OP-TEE payload + U-boot) and Linux kernel (`kernel8.img`) are stored on the boot partition. With the default configuration, there are only 2MBs of free space, after all this stuff is copied there. Not sure how  `armstub8.bin` blob extends when you enable TBB, but just to exclude the possible issues with free storage I would recommend to extend it.

BTW, let me try to setup TBB on my side and see what's is wrong

**Comment by jbech-linaro** [comment]
> Im guessing the main problem should be that `gen_rootfs` is missing or is that deprecated ?

Right that is not being used any longer, it was [replaced by Buildroot a couple of months ago](https://github.com/OP-TEE/build/pull/268). So don't spend time on trying to use gen_rootfs.

**Comment by Christian56** [comment]
Thanks for the advice :) 
I found the issue  /out-br/images/rootfs.cpio contained symbolic links, which didnt work for the filesystem, which was used (fat32, I use ext3 now). 
I think now i got an issue with udhcpc as i dont get an "OK" in the terminal. 
It stops after : 
```
...
Starting logging: OK
[    5.995213] usb 1-1.1: new high-speed USB device number 3 using dwc_otg
Initializing random number generator... [    6.046975] random: dd: uninitialized urandom read (512 bytes read)
done.
Setting permissions on /dev/tee*... OK
Starting tee-supplicant... [    6.122762] usb 1-1.1: New USB device found, idVendor=0424, idProduct=ec00
[    6.135713] usb 1-1.1: New USB device strings: Mfr=0, Product=0, SerialNumber=0
[    6.151814] smsc95xx v1.0.6
OK
Starting network: [    6.207798] smsc95xx 1-1.1:1.0 eth0: register 'smsc95xx' at usb-3f980000.usb-1.1, smsc95xx USB 2.0 Ethernet, b8:27:eb:8f:c4:dd
OK
Starting network (udhcpc): [    6.339886] smsc95xx 1-1.1:1.0 eth0: hardware isn't capable of remote wakeup
[    7.523687] random: crng init done
```
As it says "Starting logging" Im wondering if it possible see the logs even through its frozen? I looked on the SD card but couldnt find anything

**Comment by jforissier** [comment]
Hi @Christian56,

It seems that de DHCP client (udhcpc) never times out, or at least not within a reasonable delay. In fact, when I added `udhcp`to the boot sequence (in commit [751b35bb](https://github.com/OP-TEE/build/commit/751b35bb)), my intention was to add it for QEMU only, as it should be clear from the overlay path (`br-ext/board/qemu/overlay`).

So the first thing we need to do is fix this problem. Could you try the attached patch, please?

[udhcpc-qemu.txt](https://github.com/OP-TEE/optee_os/files/2512946/udhcpc-qemu.txt)

Then, if you need DHCP networking, some more troubleshooting would be needed.

**Comment by Christian56** [comment]
Hi @jforissier, 
I applied the patch and compiled the system via make all, but the issue remains the same. 

```
Starting logging: OK
[    6.009732] usb 1-1.1: New USB device found, idVendor=0424, idProduct=ec00
[    6.050988] usb 1-1.1: New USB device strings: Mfr=0, Product=0, SerialNumber=0
Initializing random number generator... [    6.091362] smsc95xx v1.0.6
[    6.099462] random: dd: uninitialized urandom read (512 bytes read)
done.
Setting permissions on /dev/tee*... [    6.161531] smsc95xx 1-1.1:1.0 eth0: register 'smsc95xx' at usb-3f980000.usb-1.1, smsc95xx USB 2.0 Ethernet, b8:27:eb:8f:c4:dd
OK
Starting tee-supplicant... OK
Starting network: OK
Starting network (udhcpc): [    6.362881] smsc95xx 1-1.1:1.0 eth0: hardware isn't capable of remote wakeup
[    7.444678] random: crng init done
```

**Comment by jforissier** [comment]
The line "Starting network (udhcpc):" should not be there anymore. I think you need to run "make buildroot-clean" first, because the patch only changes Makefiles.

**Comment by Christian56** [comment]
makes sense :+1:, i cleaned the buildroot and i ran xtest. 

```
16081 subtests of which 0 failed
74 test cases of which 0 failed
0 test case was skipped
TEE test application done!
```
Thanks a lot for your help and support

**Comment by Christian56** [comment]
Not sure where to put this, but in order to compile with the ATF Params necessary for secure boot 
   ```
MBEDTLS_DIR=path of the directory containing mbed TLS sources
    TRUSTED_BOARD_BOOT=1
    GENERATE_COT=1
    ARM_ROTPK_LOCATION=devel_rsa
    ROT_KEY=$(ARM_TF_PATH)/plat/arm/board/common/rotpk/arm_rotprivk_rsa.pem
```
The header file and the C file of platform_util and rsa_internal needs to be added to mbedtls.

**Comment by jbech-linaro** [comment]
@Christian56 , we have most of it I think already here: 
https://github.com/OP-TEE/optee_os/blob/master/documentation/optee_with_auth_framework.md

**Comment by Christian56** [comment]
Hi @jbech-linaro, 
I tried to verify the secure boot by changing a byte in armstub8.bin or fip.bin. However the boot process didnt stop nor did i get a warning. When i compiled the system i got Notice that my key have been created and when i boot up the system i got notice like that: 

```
NOTICE:  Booting Trusted Firmware
NOTICE:  BL1: v1.5(debug):v1.5-620-g8ae41ae
NOTICE:  BL1: Built : 17:36:58, Dec 10 2018
INFO:    BL1: RAM 0x100ee000 - 0x100f9000
INFO:    BL1: cortex_a53: CPU workaround for 843419 was applied
INFO:    BL1: cortex_a53: CPU workaround for 855873 was applied
NOTICE:  rpi3: Detected: Raspberry Pi 3 Model B (1GB, Sony, UK) [0x00a02082]
INFO:    BL1: Loading BL2
INFO:    Loading image id=1 at address 0x100b4000
INFO:    Image id=1 loaded: 0x100b4000 - 0x100ba431
NOTICE:  BL1: Booting BL2
INFO:    Entry point address = 0x100b4000
INFO:    SPSR = 0x3c5
NOTICE:  BL2: v1.5(debug):v1.5-620-g8ae41ae
NOTICE:  BL2: Built : 17:36:59, Dec 10 2018
```

So i assume the secure boot should be working and i need either another configuration parameter or that my changed byte is not part of the BL Stage (but armstub8.bin should be i guess).
